pkg/objectcache/applicationprofilecache/applicationprofilecache.go

```diff
@@ -55,91 +36,252 @@ type ContainerCallStackIndex struct {
-	slugToContainers    maps.SafeMap[string, mapset.Set[string]]          // cache the containerIDs that belong to the application profile, this will enable removing from cache AP without pods
-	slugToState         maps.SafeMap[string, applicationProfileState]     // cache the containerID to slug mapping, this will enable a quick lookup of the application profile
-	storageClient       versioned.SpdxV1beta1Interface
-	allProfiles         mapset.Set[string] // cache all the application profiles that are ready. this will enable removing from cache AP without pods that are running on the same node
-	userManagedProfiles maps.SafeMap[string, *v1beta1.ApplicationProfile]
-	containerCallStacks maps.SafeMap[string, *ContainerCallStackIndex] // cache the containerID to call stack search tree mapping
-	containerToName     maps.SafeMap[string, string]                   // cache the containerID to container name mapping
+	cfg                            config.Config
+	workloadIDToProfile            maps.SafeMap[string, *v1beta1.ApplicationProfile]
+	containerIDToInfo              maps.SafeMap[string, *ContainerInfo]
+	namespaceToContainers          maps.SafeMap[string, mapset.Set[string]] // namespace -> set of containerIDs
+	profileToUserManagedIdentifier maps.SafeMap[string, string]             // profileName -> user-managed profile unique identifier (This is used to prevent merging the same user-managed profile multiple times)
+	containerToCallStackIndex      maps.SafeMap[string, *ContainerCallStackIndex]
+	storageClient                  versioned.SpdxV1beta1Interface
+	k8sObjectCache                 objectcache.K8sObjectCache
+	updateInterval                 time.Duration
+	mutex                          sync.Mutex // For operations that need additional synchronization
 }
 
-func NewApplicationProfileCache(cfg config.Config, storageClient versioned.SpdxV1beta1Interface) *ApplicationProfileCacheImpl {
-	return &ApplicationProfileCacheImpl{
-		cfg:                 cfg,
-		storageClient:       storageClient,
-		containerToSlug:     maps.SafeMap[string, string]{},
-		slugToAppProfile:    maps.SafeMap[string, *v1beta1.ApplicationProfile]{},
-		slugToContainers:    maps.SafeMap[string, mapset.Set[string]]{},
-		slugToState:         maps.SafeMap[string, applicationProfileState]{},
-		allProfiles:         mapset.NewSet[string](),
-		userManagedProfiles: maps.SafeMap[string, *v1beta1.ApplicationProfile]{},
-		containerCallStacks: maps.SafeMap[string, *ContainerCallStackIndex]{},
-		containerToName:     maps.SafeMap[string, string]{},
+// NewApplicationProfileCache creates a new application profile cache with periodic updates
+func NewApplicationProfileCache(cfg config.Config, storageClient versioned.SpdxV1beta1Interface, k8sObjectCache objectcache.K8sObjectCache) *ApplicationProfileCacheImpl {
+	updateInterval := utils.AddJitter(cfg.ProfilesCacheRefreshRate, 10) // Add 10% jitter to avoid high load on the storage
```

Comment:
It would be much better to add the jitter to the global frequency (setting it per node-agent instance), but in between loops! In this case the the agent will have a 10% global frequency difference over time, I am not sure you wanted this. (think about it, if a node agent runs for 30 days, it might did 10% less updates you meant it to do)